### PR TITLE
Remove mod_constraints field of module body

### PR DESCRIPTION
**Pull request by SkySkimmer**
Overlay: https://github.com/LPCIC/coq-elpi/pull/131

### CODE REVIEW

**Comment by ppedrot** [comment]
Can you give a high-level description of this change?

**Comment by ppedrot** [comment]
I support the idea, but the implementation is broken. You're probably forgetting to push a set of constraints somewhere.

**Comment by ppedrot** [comment]
Minimal failing example:
```
Module Foo.

Definition foo : Type := Set.

End Foo.

Lemma bar : Foo.foo = Set.
```

File: kernel/mod_typing.ml
Comment (ppedrot) [inline kernel/mod_typing.ml:309]: This `cst1` shouldn't be passed as an argument.

File: kernel/safe_typing.ml
Comment (ppedrot) [inline kernel/safe_typing.ml:1198]: This is the source of the problem. The `newenv` does not contain the constraints of the module at that point, because it is the environment before the module was opened. You need to reset its universe graph here in the same way that the opaque table and native symbols are set above.

File: kernel/mod_typing.ml
Comment (SkySkimmer) [reply kernel/mod_typing.ml:309]: This is annoying to change with the way the `translation` type and code plumbing work, feel free to push a patch on top but not worth the effort for me.

**Comment by gares** [comment]
may I ask why? is it just cleanup or it pave the way to something else?

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
Not the author of the PR, but I really support this change as an important cleanup. The kernel is messing way too much with global universe constraints to be trusted, so every time we reduce the places such constraints can appear De Bruijn rolls a bit less in his grave.

**Comment by SkySkimmer** [comment]
This cleanup makes the data structures behave closer to how we actually handle universes: they are library level, not module level.
As a bonus we now statically enforce that the kernel module typechecking doesn't produce new universes (mod_typing uses constraint.t instead of contextset.t)

**Comment by ppedrot** [comment]
I'd like to have an independent look by somebody from @coq/kernel-maintainers at this PR before merging, because it's the kind of tricky code that can introduce soundness issues without realizing it.

**Comment by ppedrot** [comment]
@maximedenes maybe?

**Comment by ppedrot** [comment]
This doesn't seem to attract the crowds, so I think I'll merge this without a second opinion. Please burn me on a stake if this makes Coq more unsound than it is currently.

**Pull request merged**
